Async plot images

ajmcmiddlin opened the pull request:
Plot images in parallel to speed up generation of multiple images.

WIP because further testing by @cameronreid.

ajmcmiddlin committed:
Add plot_images to make parallel requests

ajmcmiddlin committed:
Refactor tests

jamiecook commented on lib/plotlyrb/plot_image.rb:
```diff
@@ -25,5 +25,33 @@ def plot_image(plot_image_spec, image_path)
+    # [(spec, image_path)] -> Integer -> ()
+    def self.plot_images(headers, spec_path_pairs, timeout)
+      spec_path_pairs.
+        map { |sp| AsyncJob.from_spec_path(PlotImage.new(headers), sp[0], sp[1], timeout) }.
```
isn't this just map(&:join)?

jamiecook commented on test/plotlyrb/plot_image_test.rb:
```diff
@@ -3,17 +3,36 @@
+
+    def test_plot_images
+      num_requests = 5
+      spec_paths = (1..num_requests).inject([]) { |sps, i|
```
again... just a map?

ajmcmiddlin committed:
Make better after Jamie's feedback

cameronreid commented:
This seems to be working pretty nicely! As we speak I'm getting 2s per image as opposed to ~5s. That's in batches of five but I'm also going to be testing batches of up to 200 and seeing what size is the best.

One thing that we could potentially implement is this - in any one job in OT, the very first request to Plotly takes a lot longer than the rest. This is okay in the original use case, but in parallel you multiply this extra delay out by however many you're batching. 

For example, when you do one at a time my first image takes 30s. When doing 5 at a time it takes 90s to get those first five and 10s for each batch of five thereafter. 

Perhaps if we could do the very first image separately then we would see some half-decent speed increases for medium amounts of requests. Of course, this is going to require all kinds of horribleness with global variables or something so it might not be worth it.

ajmcmiddlin commented:
Weird. Simple test image (~3 point scatter) takes ~2s. If this 'priming' effect is happening, you could try starting with a dead simple plot and discarding it I guess. That _should_ just be a workaround though - longer term we should try and find out why this is happening and address it.

cameronreid commented:
So I haven't had any major issues with this. I'm having a fair bit of success with 8 images at a time. But, some kind of test (in both the asynchronous and regular versions) which would retry on failure (say, if the outputted image is less than a few kilobytes) would be incredibly useful. With 8 at a time I'm getting a success rate of about 95% and it would be awesome to bring it up to 100%.

ajmcmiddlin commented:
I didn't want to bake the retry logic into this, however it does return an [`AsyncJobResult`](https://github.com/vlc/plotlyrb/pull/3/files#diff-6d9712c497a5b1fe05ff260a88da23efR29) for each request. That gives you back a boolean for success, as well as the spec and path for that image, so when you get your results back you can do something like...

```.ruby
rework = results.select(&:success).map { |r| [r.spec, r.path] }
plot_images(redo, timeout)
```

Maybe we should just add `num_retries` as an argument and do this for the user. Especially if failures are so common.

ajmcmiddlin commented:
That `select` should be a `reject`. My bad!

ajmcmiddlin committed:
Make better

 - Use SpecPath struct everywhere to keep consistent
 - Exit failed threads in an attempt to cleanup

ajmcmiddlin committed:
Rejigger failed_spec_paths and test

ajmcmiddlin committed:
Add retry logic to plot_images

ajmcmiddlin commented:
@cameronreid Have a crack at this. The interface has changed a little. `plot_images` takes an extra parameter, which is the number of retries to attempt if there are any failures. The return value is now just the list of failures (after all retries) - not all results.

cameronreid committed:
Recursive retry method

cameronreid commented on lib/plotlyrb/plot_image.rb:
```diff
@@ -38,25 +38,32 @@ def self.from_spec_path(pi, spec_path, timeout)
+        AsyncJob.from_spec_path(PlotImage.new(headers), sp, timeout)
+      }.map(&:join)
+      failed_paths = failed_spec_paths(async_job_results)
+      plot_images(headers, failed_paths, timeout, retries, attempt + 1)
```
@ajmccluskey this is a pretty big change but I was messing around with why everything was retrying and have rewritten it a bit. It may have been failing because the threads always return nil or something? In any case, I haven't had any issues with this implementation. Your thoughts?

ajmcmiddlin committed:
Refine check of success

 - Use Response class to detect API failures from Plotly
 - Bubble exceptions from threads as AsyncJobResults

ajmcmiddlin committed:
Make working, fix test

ajmcmiddlin committed:
Fix joining of AsyncJobs - that was bad

ajmcmiddlin committed:
Change plot_images so it returns AsyncJobResults

ajmcmiddlin committed:
Fix comment

ajmcmiddlin commented:
@cameronreid there were definitely some issues there. I was actually in the process of trying to clean it up myself, so I've mostly continued on that path.

I added the file existence check from your code, however I removed the size check, as I don't think we should rely on that. Instead I've used the `Response` object I created for grid and plot creation. `Response` checks the HTTP response code from Plotly, which should be set correctly when there are errors.

I've also changed the interface again - now `plot_images` returns a list of `AsyncJobResults` for failures. `AsyncJobResult` contains the `SpecPath` used in the request, as well as a message that you can use to log failures.

Finally, the reason the old code was weird was that it was retrying ALL of your inputs when there were one or more failures. I did a bad and used the wrong variable to launch requests.

Give this updated code a burl - hopefully it works as well as yours. Let me know if it doesn't and we can work through it.

ajmcmiddlin committed:
Remove unused var

ajmcmiddlin commented:
@cameronreid are you OK with this version? @jamiecook this has changed quite significantly - care to cast your eyes over the latest?

cameronreid commented:
I'm cool with this.
